File: neo_todo/task_manager.py
"""Task management module implementing the 3-Column Lifecycle.

Zones:
- Backpack (Plan): Tasks planned for the day with a capacity limit.
- Court (Focus): Single-task zone where active work happens.
- Completed (Trophies): Finished tasks that contribute to the efficiency index.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import IntEnum
from typing import Optional
from uuid import uuid4


class TaskDifficulty(IntEnum):
    """Task difficulty levels with associated point values."""

    EASY = 10
    MEDIUM = 30
    HARD = 60
    EPIC = 100


@dataclass
class Task:
    """Represents a single task in the system."""

    title: str
    difficulty: TaskDifficulty
    id: str = field(default_factory=lambda: str(uuid4()))
    elapsed_minutes: float = 0.0

    @property
    def points(self) -> int:
        """Point value derived from difficulty."""
        return int(self.difficulty)


class TaskManager:
    """Manages the 3-column task lifecycle.

    Enforces:
    - Backpack capacity limit (max total points per day).
    - Single-task rule on the court.
    - Ordered flow: Backpack -> Court -> Completed (or Backpack <- Court).
    """

    DEFAULT_BACKPACK_CAPACITY = 300

    def __init__(self, backpack_capacity: int = DEFAULT_BACKPACK_CAPACITY) -> None:
        self._backpack_capacity = backpack_capacity
        self._backpack: list[Task] = []
        self._court: Optional[Task] = None
        self._completed: list[Task] = []

    @property
    def backpack_capacity(self) -> int:
        """Maximum total points allowed in the backpack."""
        return self._backpack_capacity

    @property
    def backpack(self) -> list[Task]:
        """Tasks in the backpack (plan) zone."""
        return list(self._backpack)

    @property
    def court_task(self) -> Optional[Task]:
        """The single task currently on the court (focus zone)."""
        return self._court

    @property
    def completed(self) -> list[Task]:
        """Tasks in the completed (trophies) zone."""
        return list(self._completed)

    @property
    def backpack_points_used(self) -> int:
        """Total points currently allocated in the backpack."""
        return sum(t.points for t in self._backpack)

    @property
    def backpack_points_remaining(self) -> int:
        """Points still available in the backpack."""
        return self._backpack_capacity - self.backpack_points_used

    @property
    def total_completed_points(self) -> int:
        """Sum of points from all completed tasks."""
        return sum(t.points for t in self._completed)

    def add_to_backpack(self, task: Task) -> None:
        """Add a task to the backpack zone.

        Args:
            task: The task to add.

        Raises:
            ValueError: If adding the task would exceed backpack capacity.
        """
        if self.backpack_points_used + task.points > self._backpack_capacity:
            raise ValueError(
                f"Cannot add task ({task.points} pts): "
                f"would exceed backpack capacity of {self._backpack_capacity} pts "
                f"(currently {self.backpack_points_used} pts used)"
            )
        self._backpack.append(task)

    def move_to_court(self, task_id: str) -> Task:
        """Move a task from the backpack to the court (focus zone).

        Args:
            task_id: ID of the task to move.

        Returns:
            The task that was moved to the court.

        Raises:
            ValueError: If court is already occupied or task not found.
        """
        if self._court is not None:
            raise ValueError(
                "Court is occupied. Complete or return the current task first."
            )

        task = self._find_and_remove_from_backpack(task_id)
        self._court = task
        return task

    def complete_court_task(self) -> Task:
        """Move the current court task to completed zone.

        Returns:
            The completed task.

        Raises:
            ValueError: If no task is on the court.
        """
        if self._court is None:
            raise ValueError("No task on the court to complete.")

        task = self._court
        self._court = None
        self._completed.append(task)
        return task

    def return_to_backpack(self) -> Task:
        """Return the court task back to the backpack.

        Returns:
            The task that was returned.

        Raises:
            ValueError: If no task is on the court.
        """
        if self._court is None:
            raise ValueError("No task on the court to return.")

        task = self._court
        self._court = None
        self._backpack.append(task)
        return task

    def remove_from_backpack(self, task_id: str) -> Task:
        """Remove a task from the backpack entirely.

        Args:
            task_id: ID of the task to remove.

        Returns:
            The removed task.

        Raises:
            ValueError: If task not found in backpack.
        """
        return self._find_and_remove_from_backpack(task_id)

    def _find_and_remove_from_backpack(self, task_id: str) -> Task:
        """Find and remove a task from the backpack by ID.

        Raises:
            ValueError: If task not found.
        """
        for i, task in enumerate(self._backpack):
            if task.id == task_id:
                return self._backpack.pop(i)
        raise ValueError(f"Task with id '{task_id}' not found in backpack.")

File: src/taskManager.js
/**
 * Task Manager — 3-Column Lifecycle.
 *
 * Zones: Backpack (plan) → Now (focus) → Done (completed)
 * Enforces single-task rule on Now and capacity limit on Backpack.
 */

import { randomUUID } from 'crypto';

/** Valid point values for tasks. */
export const TASK_POINTS = { SMALL: 10, MEDIUM: 50, LARGE: 100 };

/** Default daily capacity for the Backpack zone. */
export const DEFAULT_CAPACITY = 300;

/**
 * Create a new task object.
 * @param {string} title
 * @param {number} points - Must be 10, 50, or 100.
 * @returns {object}
 */
export function createTask(title, points) {
  const validPoints = [TASK_POINTS.SMALL, TASK_POINTS.MEDIUM, TASK_POINTS.LARGE];
  if (!validPoints.includes(points)) {
    throw new Error(`Invalid points value: ${points}. Must be one of ${validPoints.join(', ')}`);
  }
  if (!title || typeof title !== 'string') {
    throw new Error('Title must be a non-empty string');
  }
  return {
    id: randomUUID(),
    title,
    points,
    status: 'backpack',
    timerStart: null,
    timerElapsedSeconds: 0,
    completedAt: null,
  };
}

export class TaskManager {
  /**
   * @param {number} [capacity=DEFAULT_CAPACITY] - Max points in Backpack.
   */
  constructor(capacity = DEFAULT_CAPACITY) {
    this._capacity = capacity;
    this._backpack = [];
    this._now = null;
    this._done = [];
  }

  get capacity() {
    return this._capacity;
  }

  get backpack() {
    return [...this._backpack];
  }

  get nowTask() {
    return this._now;
  }

  get done() {
    return [...this._done];
  }

  /** Total points currently in the backpack. */
  get backpackPointsUsed() {
    return this._backpack.reduce((sum, t) => sum + t.points, 0);
  }

  /** Points remaining before capacity is reached. */
  get backpackPointsRemaining() {
    return this._capacity - this.backpackPointsUsed;
  }

  /** Total points earned from completed tasks. */
  get totalPointsEarned() {
    return this._done.reduce((sum, t) => sum + t.points, 0);
  }

  /**
   * Add a task to the Backpack.
   * @param {object} task - Created via createTask().
   */
  addToBackpack(task) {
    if (this.backpackPointsUsed + task.points > this._capacity) {
      throw new Error(
        `Cannot add task (${task.points} pts): would exceed capacity of ${this._capacity} pts ` +
        `(${this.backpackPointsUsed} pts used)`
      );
    }
    task.status = 'backpack';
    this._backpack.push(task);
  }

  /**
   * Move a task from Backpack to Now (active focus).
   * @param {string} taskId
   * @param {number} [timestamp] - Optional ISO timestamp or epoch ms for timer start.
   * @returns {object} The task moved to Now.
   */
  moveToNow(taskId, timestamp = Date.now()) {
    if (this._now !== null) {
      throw new Error('Now zone is occupied. Complete or return the current task first.');
    }
    const idx = this._backpack.findIndex(t => t.id === taskId);
    if (idx === -1) {
      throw new Error(`Task '${taskId}' not found in Backpack.`);
    }
    const [task] = this._backpack.splice(idx, 1);
    task.status = 'now';
    task.timerStart = timestamp;
    this._now = task;
    return task;
  }

  /**
   * Complete the current Now task and move it to Done.
   * @param {number} [timestamp] - Optional timestamp for completion.
   * @returns {object} The completed task.
   */
  completeNowTask(timestamp = Date.now()) {
    if (this._now === null) {
      throw new Error('No task in Now zone to complete.');
    }
    const task = this._now;
    if (task.timerStart !== null) {
      task.timerElapsedSeconds += (timestamp - task.timerStart) / 1000;
    }
    task.timerStart = null;
    task.status = 'done';
    task.completedAt = timestamp;
    this._now = null;
    this._done.push(task);
    return task;
  }

  /**
   * Return the current Now task back to the Backpack.
   * @param {number} [timestamp] - Optional timestamp.
   * @returns {object} The returned task.
   */
  returnToBackpack(timestamp = Date.now()) {
    if (this._now === null) {
      throw new Error('No task in Now zone to return.');
    }
    const task = this._now;
    if (task.timerStart !== null) {
      task.timerElapsedSeconds += (timestamp - task.timerStart) / 1000;
    }
    task.timerStart = null;
    task.status = 'backpack';
    this._now = null;
    this._backpack.push(task);
    return task;
  }

  /**
   * Remove a task from the Backpack entirely.
   * @param {string} taskId
   * @returns {object} The removed task.
   */
  removeFromBackpack(taskId) {
    const idx = this._backpack.findIndex(t => t.id === taskId);
    if (idx === -1) {
      throw new Error(`Task '${taskId}' not found in Backpack.`);
    }
    return this._backpack.splice(idx, 1)[0];
  }
}
